### ISSUE TITLE
VPA Updater stuck in POD eviction loop

### ISSUE THREAD

**Issue by CezarManea** [kubernetes/autoscaler#4774]
**Which component are you using?**:
VPA

**What version of the component are you using?**:
0.10.0

Component version:

**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-02-16T12:38:05Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"darwin/arm64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.8", GitCommit:"4a3b558c52eb6995b3c5c1db5e54111bd0645a64", GitTreeState:"clean", BuildDate:"2021-12-15T14:46:22Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
</pre></details>

**What environment is this in?**:
On premise

**What did you expect to happen?**:
If container of a pod is outside recommended range to be evicted once. The newly created pod to have the recommended resources request and wouldn't be evicted for the next 24h.

**What happened instead?**:
Pod is evicted by **VPA updater** every 1 minute. The newly created pod has "almost" the recommended resources request set.
I was able to debug the issue and find out that precision calculation in different VPA components (recommender/admission controller) is causing the issue. 

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a VPA object with the following specs (the minAllowed cpu value is very important)
```
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: eviction-loop-vpa
  namespace: default
spec:
  resourcePolicy:
    containerPolicies:
    - containerName: nginx
      maxAllowed:
        cpu: 375m
        memory: 768Mi
      minAllowed:
        cpu: 187500u
        memory: 0.6Gi
  targetRef:
    apiVersion: apps/v1
    kind: StatefulSet
    name: eviction-loop-vpa
  updatePolicy:
    updateMode: Auto
```
2.  Wait a couple of hours for the recommender to make confident recommendations
```
status:
  conditions:
  - lastTransitionTime: "2022-03-29T14:05:49Z"
    status: "True"
    type: RecommendationProvided
  recommendation:
    containerRecommendations:
    - containerName: nginx
      lowerBound:
        cpu: 187500u
        memory: 563714457600m
      target:
        cpu: 187500u
        memory: 563714457600m
      uncappedTarget:
        cpu: 109m
        memory: "272061154"
      upperBound:
        cpu: 187500u
        memory: 563714457600m
```
    
3. The pod is evicted almost every minute and recreated with the following resources requests
```
    resources:
      limits:
        cpu: 250m
        memory: "1127428916"
      requests:
        cpu: 188m
        memory: 563714457600m
```
4. Because of this [check](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/priority/priority_processor.go#L73) the request (188m) is always grater than upperBound (187500u) and the pod will be evicted. 
The newly created pod will have same resources set as the previous one. The pod is again evicted because of the check and so on.

**Anything else we need to know?**:
The issue is caused by the resources set by the **vpa-admission-controller** (187500u and not 188m), but I think it also could be fixed by setting a more permissive precision in all VPA components.

**Comment by jbartosik** [comment]
I suspect this may be the same issue as https://github.com/kubernetes/autoscaler/issues/4733.

Please check if your webhook is using v1 API. You can do this by running:

kubectl get MutatingWebhookConfiguration vpa-webhook-config -oyaml

If the webhook is using v1beta2 API then you should switch to using v1. When Admission Controller starts it by default deletes existing webhook and creates a new one (properly configured). So if you have v1beta2 webhook I think there are two possibilities:

You're running VPA with [registerWebhook](https://github.com/kubernetes/autoscaler/blob/f93d361836dcbe4788a115da981a7ebf940ed337/vertical-pod-autoscaler/pkg/admission-controller/main.go#L68) flag set to false
Something is changing webhook after VPA Admission Controller updates it.
Please let me know if this helps.

**Comment by CezarManea** [comment]
I'm using v1 API for MutatingWebhookConfiguration
```
❯ kubectl get MutatingWebhookConfiguration vpa-webhook-config -oyaml
apiVersion: admissionregistration.k8s.io/v1
```

Also, **register-webhook** is no set to false, therefore using the true default value.

I have implemented a workaround for this issue, always round CPU resources to milicores and memory to Mi.
With hundreds of VPA objects I haven't seen any POD eviction loop.

Is there anything else I can check to see if the issue is my configuration?

**Comment by jbartosik** [comment]
I think you should check a different field, when I reproduced the problem the webhook looked like that:

```
apiVersion: admissionregistration.k8s.io/v1
kind: MutatingWebhookConfiguration
metadata:
  name: vpa-webhook-config
webhooks:
- admissionReviewVersions:
  - v1beta1
```

So `v1beta1` was a key in `webhooks.admissionReviewVersions`

**Comment by CezarManea** [comment]
Here's the more detailed snippet
```
❯ kubectl get MutatingWebhookConfiguration vpa-webhook-config -o yaml
apiVersion: admissionregistration.k8s.io/v1
kind: MutatingWebhookConfiguration
metadata:
  creationTimestamp: "2022-03-29T13:24:41Z"
  generation: 1
  name: vpa-webhook-config
  resourceVersion: "3485783384"
  uid: b059e9b0-b52f-4d9c-9dd4-27d4f7257bc6
webhooks:
- admissionReviewVersions:
  - v1
  clientConfig:
....
```

**Comment by jbartosik** [comment]
Thanks for checking. We have [an e2 etest](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-0.10.0/vertical-pod-autoscaler/e2e/v1/admission_controller.go#L408) which verifies that VPA will cap recommendation of 250 mCPU to `MaxAllowed` of 233mCPU and the test was passing when I cut the release so I wanted to make sure that Admission Controller is working.

When I was trying to reproduce the issue I noticed that `minAllowed` you specified (`187500u`) has fraction of mCPU. The problem with that is that "Kubernetes doesn't allow you to specify CPU resources with a precision finer than 1m." ([documentation](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu)). (I tried to create a pod with `188500u` CPU and got a pod wit h `189m`.

I think it would be good to add a validation and reject VPA objects with resource specifications that go below supported precision.
